src/traces/pie/plot.js

```diff
@@ -454,6 +501,72 @@ function transformOutsideText(textBB, pt) {
     };
 }
 
+function positionTitleInside(titleBB, cd0) {
+    var textDiameter = Math.sqrt(titleBB.width * titleBB.width + titleBB.height * titleBB.height);
+    return {
+        x: cd0.cx,
+        y: cd0.cy,
+        scale: cd0.trace.hole * cd0.r * 2 / textDiameter
+    };
+}
+
+function positionTitleOutside(titleBB, cd0, plotSize) {
+    var scaleX, scaleY, chartWidth, titleSpace, titleShift, maxPull;
+    var trace = cd0.trace;
+
+    maxPull = getMaxPull(trace);
+    chartWidth = plotSize.w * (trace.domain.x[1] - trace.domain.x[0]);
+    scaleX = chartWidth / titleBB.width;
+    if(isSinglePie(trace)) {
+        titleShift = trace.titlefont.size / 2;
+        // we need to leave enough free space for an outside label
+        if(trace.outsidetextfont) titleShift += 3 * trace.outsidetextfont.size / 2;
+        else titleShift += trace.titlefont.size / 4;
+        return {
+            x: cd0.cx,
+            y: cd0.cy - (1 + maxPull) * cd0.r - titleShift,
+            scale: scaleX
+        };
+    }
+    titleSpace = getTitleSpace(trace, plotSize);
+    // we previously left a free space of height titleSpace.
+    // The text must fit in this space.
+    scaleY = titleSpace / titleBB.height;
+    return {
+        x: cd0.cx,
+        y: cd0.cy - (1 + maxPull) * cd0.r - (titleSpace / 2),
+        scale: Math.min(scaleX, scaleY)
+    };
+}
+
+function isSinglePie(trace) {
+    // check if there is a single pie per y-column
+    if(trace.domain.y[0] === 0 && trace.domain.y[1] === 1) return true;
+    return false;
+}
+
+function getTitleSpace(trace, plotSize) {
+    var chartHeight = plotSize.h * (trace.domain.y[1] - trace.domain.y[0]);
+    // leave 3/2 * titlefont.size free space. We need at least titlefont.size
+    // space, and the 1/2 * titlefont.size is a small buffer to avoid the text
+    // touching the pie.
```

Comment:
This should be based on `titleBB.height` - since you can include `'multi<br>line<br>titles'`
But a buffer of half the font size is still good.

Actually, in principle anything that depends on the size of text needs to happen in the (potentially async) callback 3rd argument to `convertToTspans`, in order to handle MathJax (`title: '$x+y=z$'` etc). That may throw a bit of a wrench into this whole process, since `scalePies` happens so early on in the process... seems like perhaps all the pie titles need to be rendered upfront, stashed somewhere, and *all* the existing plotting happens in their combined completion callback. Super annoying, I know...

Reply:
I handled multiline titles.
Making it work for MathJax is much harder. Maybe I could just use `data-notex`?

Reply:
per https://github.com/plotly/plotly.js/pull/2987#issuecomment-421176694 - yes, just use `data-notex`.